### PR TITLE
MM-27120 - Document how to use external database and filestore with Mattermost Operator

### PR DESCRIPTION
**Pull request by Szymongib**
#### Summary
Document how to use external database and filestore with Installations created with Mattermost Operator.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-27120

### CODE REVIEW

File: source/install/install-kubernetes-mattermost.rst [hunk 1]
```diff
@@ -76,6 +76,8 @@ Open a text editor and create a text file with the following details.
 
 Save the file as ``mattermost-installation.yaml``.
 
+To configure the Mattermost Installation with external database and filestore refere to `Configuring external database and filestore (Recommended)`_ before proceeding.
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:79]:
```suggestion
To configure the Mattermost Installation with an external database and filestore refer to `Configuring external database and filestore (Recommended)`_ before proceeding.
```

There's a link missing here - is that intentional or is it a reference to another section of the document?

Comment (Szymongib) [reply source/install/install-kubernetes-mattermost.rst:79]: Yes, it is a link to another section of the document.

File: source/install/install-kubernetes-mattermost.rst [hunk 2]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
 
 Navigate to the ``ingressName`` URL in your browser and use Mattermost.
 
+
+Configuring external database and filestore (Recommended)
+----------------------------------------------------------
+
+When installing Mattermost with Mattermost Operator for production scenario it is recommended to use an external database and filestore.
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:122]:
```suggestion
When installing Mattermost using the Mattermost Operator in a production scenario, it's recommended that you use an external database and filestore.
```

File: source/install/install-kubernetes-mattermost.rst [hunk 3]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+**1. Create database Secret**
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:124]:
```suggestion
1. Create database secret
^^^^^^^^^^^^^^^^^^^
```

File: source/install/install-kubernetes-mattermost.rst [hunk 4]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+The database Secret needs to be created in the namespace that will hold the Mattermost Installation.
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:126]:
```suggestion
The database secret needs to be created in the namespace that will hold the Mattermost installation.
```

File: source/install/install-kubernetes-mattermost.rst [hunk 5]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+The Secret should contain the following data:
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:127]:
```suggestion
The secret should contain the following data:
```

File: source/install/install-kubernetes-mattermost.rst [hunk 6]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+.. csv-table::
+    :header: "Key", "Description", "Required"
+
+    "DB_CONNECTION_STRING", "Connection string to the database.", "true"
+    "MM_SQLSETTINGS_DATASOURCEREPLICAS", "Connection string to read replicas of the database.", "false"
+    "DB_CONNECTION_CHECK_URL", "The URL used for checking that the database is accessible.", "false"
+
+Example Secret for AWS Aurora compatible with PostgreSQL:
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:136]:
```suggestion
Example secret for AWS Aurora compatible with PostgreSQL:
```

File: source/install/install-kubernetes-mattermost.rst [hunk 7]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+    DB_CONNECTION_CHECK_URL: cG9zdGdyZXM6Ly91c2VyOnN1cGVyX3NlY3JldF9wYXNzd29yZEBteS1kYXRhYmFzZS5jbHVzdGVyLWFiY2QudXMtZWFzdC0xLnJkcy5hbWF6b25hd3MuY29tOjU0MzIvbWF0dGVybW9zdD9jb25uZWN0X3RpbWVvdXQ9MTAK
+    DB_CONNECTION_STRING: cG9zdGdyZXM6Ly91c2VyOnN1cGVyX3NlY3JldF9wYXNzd29yZEBteS1kYXRhYmFzZS5jbHVzdGVyLWFiY2QudXMtZWFzdC0xLnJkcy5hbWF6b25hd3MuY29tOjU0MzIvbWF0dGVybW9zdD9jb25uZWN0X3RpbWVvdXQ9MTAK
+    MM_SQLSETTINGS_DATASOURCEREPLICAS: cG9zdGdyZXM6Ly91c2VyOnN1cGVyX3NlY3JldF9wYXNzd29yZEBteS1kYXRhYmFzZS5jbHVzdGVyLXJvLWFiY2QudXMtZWFzdC0xLnJkcy5hbWF6b25hd3MuY29tOjU0MzIvbWF0dGVybW9zdD9jb25uZWN0X3RpbWVvdXQ9MTAK
+  kind: Secret
+  metadata:
+    name: my-postgres-connection
+  type: Opaque
+
+.. note:: For PostgreSQL the connection is checked with `pg_isready <https://www.postgresql.org/docs/9.3/app-pg-isready.html>`__ so the ``DB_CONNECTION_CHECK_URL`` is the same as connection string.
+   For MySQL the check is performed via HTTP call therefore ``DB_CONNECTION_CHECK_URL`` should be an HTTP url.
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:151]:
```suggestion
   For MySQL the check is performed via HTTP call therefore ``DB_CONNECTION_CHECK_URL`` should be an HTTP URL.
```

File: source/install/install-kubernetes-mattermost.rst [hunk 8]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+
+**2. Create filestore Secret**
+
+The filestore Secret needs to be created in the namespace that will hold the Mattermost Installation.
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:156]:
```suggestion
The filestore secret needs to be created in the namespace that will hold the Mattermost installation.
```

File: source/install/install-kubernetes-mattermost.rst [hunk 9]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+The Secret should contain the following data:
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:157]:
```suggestion
The secret should contain the following data:
```

File: source/install/install-kubernetes-mattermost.rst [hunk 10]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+.. csv-table::
+    :header: "Key", "Description", "Required"
+
+    "accesskey", "Filestore access key.", true
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:162]:
```suggestion
    "accesskey", "Filestore access key.", "true"
```

File: source/install/install-kubernetes-mattermost.rst [hunk 11]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+    "accesskey", "Filestore secret key.", true
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:163]:
```suggestion
    "accesskey", "Filestore secret key.", "true"
```

File: source/install/install-kubernetes-mattermost.rst [hunk 12]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+Example Secret for AWS S3:
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:165]:
```suggestion
Example secret for AWS S3:
```

File: source/install/install-kubernetes-mattermost.rst [hunk 13]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+  data:
+    accesskey: QUNDRVNTX0tFWQo=
+    secretkey: U1VQRVJfU0VDUkVUX0tFWQo=
+  kind: Secret
+  metadata:
+    name: my-s3-iam-access-key
+  type: Opaque
+
+
+**3. Adjust Cluster Installation**
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:179]:
```suggestion
3. Adjust cluster installation
^^^^^^^^^^^^^^^^^^^^^
```

File: source/install/install-kubernetes-mattermost.rst [hunk 14]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+.. code-block:: yaml
+
+  spec:
+  ...
+      mattermostEnv:
+      ...
+      - name: MM_FILESETTINGS_AMAZONS3SSE
+        value: "true"
+
+Example Cluster Installation configured with both external databases and filestore:
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:213]:
```suggestion
Example cluster installation configured with both external databases and filestore:
```

File: source/install/install-kubernetes-mattermost.rst [hunk 15]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+    minio:
+      externalBucket: my-s3-bucket
+      externalURL: s3.amazonaws.com
+      secret: my-s3-iam-access-key
+      storageSize: 50Gi
+    elasticSearch:
+      host: ""
+      username: ""
+      password: ""
+
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:240]:
```suggestion
```

**Review by justinegeffen — COMMENTED** [review]
Made some minor editorial changes and some formatting changes but otherwise LGTM. Let me know when you've reviewed the suggestions and I'll give it a second pass to validate formatting.

File: source/install/install-kubernetes-mattermost.rst [hunk 16]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+The database Secret needs to be created in the namespace that will hold the Mattermost Installation.
+The Secret should contain the following data:
+
+.. csv-table::
+    :header: "Key", "Description", "Required"
+
+    "DB_CONNECTION_STRING", "Connection string to the database.", "true"
+    "MM_SQLSETTINGS_DATASOURCEREPLICAS", "Connection string to read replicas of the database.", "false"
+    "DB_CONNECTION_CHECK_URL", "The URL used for checking that the database is accessible.", "false"
```
Comment (jwilander) [inline source/install/install-kubernetes-mattermost.rst:134]: Might be worth adding a note here that `DB_CONNECTION_CHECK_URL` is often the same as `DB_CONNECTION_STRING`

File: source/install/install-kubernetes-mattermost.rst [hunk 17]
```diff
@@ -113,6 +115,130 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+.. code-block:: yaml
+
+  spec:
+  ...
+    minio:
+      externalBucket: my-s3-bucket
+      externalURL: s3.amazonaws.com
+      secret: my-s3-iam-access-key
+
+Additionaly when using Amazon S3, set the ``MM_FILESETTINGS_AMAZONS3SSE`` environment variable to ``true``:
```
Comment (jwilander) [inline source/install/install-kubernetes-mattermost.rst:202]: Might also want to mention setting `MM_FILESETTINGS_AMAZONS3SSL` to true as well for any S3 users

**Review by jwilander — COMMENTED** [review]
Good stuff 👍 Just a couple small suggestions

File: source/install/install-kubernetes-mattermost.rst (hunk 16 above)
Comment (Szymongib) [reply source/install/install-kubernetes-mattermost.rst:134]: There is a note below example secret https://github.com/mattermost/docs/pull/4113/files#diff-ac678edad9006c078b03da9e586c48bd39bb15913a0a56e1d6f1852846148ce4R150
Not sure if you have missed it or you prefer to add this information to the table?

File: source/install/install-kubernetes-mattermost.rst [hunk 18]
```diff
@@ -113,6 +115,131 @@ pointing to the address you just copied. For example, on AWS you would do this w
 
 Navigate to the ``ingressName`` URL in your browser and use Mattermost.
 
+
+Configuring external database and filestore (Recommended)
+----------------------------------------------------------
+
+When installing Mattermost using the Mattermost Operator in a production scenario, it's recommended that you use an external database and filestore.
+
+**1. Create database secret**
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:124]:
```suggestion
1. Create database secret
^^^^^^^^^^^^^^^^^^^
```

File: source/install/install-kubernetes-mattermost.rst [hunk 19]
```diff
@@ -113,6 +115,131 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+
+  apiVersion: v1
+  data:
+    accesskey: QUNDRVNTX0tFWQo=
+    secretkey: U1VQRVJfU0VDUkVUX0tFWQo=
+  kind: Secret
+  metadata:
+    name: my-s3-iam-access-key
+  type: Opaque
+
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:177]:
```suggestion
```

File: source/install/install-kubernetes-mattermost.rst [hunk 20]
```diff
@@ -76,6 +76,8 @@ Open a text editor and create a text file with the following details.
 
 Save the file as ``mattermost-installation.yaml``.
 
+To configure the Mattermost Installation with an external database and filestore refer to `Configuring external database and filestore (Recommended)`_ before proceeding.
+
 **3. Apply the Installation Manifest File**
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply source/install/install-kubernetes-mattermost.rst:81]: This is from a different list (not modified by this PR) where all points are formated as this one, should we change them all?

File: source/install/install-kubernetes-mattermost.rst [hunk 21]
```diff
@@ -113,6 +115,131 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
 Navigate to the ``ingressName`` URL in your browser and use Mattermost.
 
+
+Configuring external database and filestore (Recommended)
+----------------------------------------------------------
+
+When installing Mattermost using the Mattermost Operator in a production scenario, it's recommended that you use an external database and filestore.
+
+1. Create database secret
+^^^^^^^^^^^^^^^^^^^
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:125]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^
```

File: source/install/install-kubernetes-mattermost.rst [hunk 22]
```diff
@@ -113,6 +115,131 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+  kind: Secret
+  metadata:
+    name: my-postgres-connection
+  type: Opaque
+
+.. note:: For PostgreSQL the connection is checked with `pg_isready <https://www.postgresql.org/docs/9.3/app-pg-isready.html>`__ so the ``DB_CONNECTION_CHECK_URL`` is the same as connection string.
+   For MySQL the check is performed via HTTP call therefore ``DB_CONNECTION_CHECK_URL`` should be an HTTP URL.
+
+2. Create filestore secret
+^^^^^^^^^^^^^^^^^^^
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:155]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: source/install/install-kubernetes-mattermost.rst [hunk 23]
```diff
@@ -113,6 +115,131 @@ pointing to the address you just copied. For example, on AWS you would do this w
...
+  data:
+    accesskey: QUNDRVNTX0tFWQo=
+    secretkey: U1VQRVJfU0VDUkVUX0tFWQo=
+  kind: Secret
+  metadata:
+    name: my-s3-iam-access-key
+  type: Opaque
+
+3. Adjust cluster installation
+^^^^^^^^^^^^^^^^^^^^^
```
Comment (justinegeffen) [inline source/install/install-kubernetes-mattermost.rst:180]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by justinegeffen — APPROVED** [review]

File: source/install/install-kubernetes-mattermost.rst (hunk 16 above)
Comment (jwilander) [reply source/install/install-kubernetes-mattermost.rst:131]: Oh yeah I missed that 👍

**Review by jwilander — APPROVED** [review]
Nice work 👍

**Pull request merged**